File: python/p0014.py
'''
Longest Collatz sequence

https://projecteuler.net/problem=14
'''


def sol():
    save = {1 : 1}
    for i in range(1000000, 0, -1):
        collatz(save, i)
    return max(save, key=save.get)


def collatz(save, n):
    if n not in save:
        if n % 2 == 0:
            save[n] = collatz(save, n // 2) + 1
        else:
            save[n] = collatz(save, 3 * n + 1) + 1
    return save[n]


if __name__ == "__main__":
    print(sol())

File: elixir/p0014.ex
defmodule P0014 do
  @moduledoc """
  Longest Collatz Sequence
  Problem 14

  https://projecteuler.net/problem=14
  """

  def solve() do
    1..1_000_000
    |> Enum.map(fn x -> {x, length(collatz_sequence(x, [x]))} end)
    |> Enum.max_by(fn {_x, y} -> y end)
    |> elem(0)
  end

  def collatz_sequence(x, history) when rem(x, 2) == 0 do
    next = div(x, 2)
    collatz_sequence(next, [next | history])
  end

  def collatz_sequence(x, history) when x > 1 and rem(x, 2) == 1 do
    next = 3 * x + 1
    collatz_sequence(next, [next | history])
  end

  def collatz_sequence(x, history) when x == 1 do
    history
  end
end

P0014.solve()
|> IO.puts()
